[Milight] Add ibox (milight v6 protocol) support and use a send queue for commands

davidgraeff opened the pull request:
This pull request adds full iBox support (Milight protocol v6).

It also refactors the way of how commands are send. The main thread is no longer blocked but the commands are added to a send queue instead with a configurable repeat time and delay between commands.

Each bulb is modeled as a separate thing now to only provide the supported channels. Instead of big switch cases each bulb is depicted as a separate class in a class hierarchy which makes the code much more readable and improve maintenance.

A session manager class is added, as the V6 bridge uses a stateful connection now.

An emulated V6 bridge helps to test the code and is able to communicate with the original app to quickly catch up with extensions made to the app at a later time.

This discussion and test results of the community can be found here: https://community.openhab.org/t/the-new-milight-ibox-and-milight-binding/17621/15?u=david_graeff

Thanks @MikeJMajor for helping to identify problems in the code. As soon as you are happy the code can be merged.

Cheers, David

MikeJMajor commented:
Perhaps a subject for another pull request, but now might be a good time considering the amount of code refactoring going on.

It would be a good idea to ensure there are 100ms pauses between commands sent to different bulbs. Sending commands to more than one channel at the same time is unreliable for me and I find I have to add sleeps in my rules.

Also other implementations (e.g. https://github.com/ojarva/python-ledcontroller/blob/master/ledcontroller/__init__.py) seem to repeat certain commands at least 3 times to improve the reliability. I find I need to do that in my rules, buts it's slow because of all the 100ms sleeps, i.e. it would be faster if the binding could send 3xON + wait 100ms + 3xSetColour.

davidgraeff commented:
@MikeJMajor I have implemented the configurable delay/repeat times.

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/QueuedCategorizedSend.java:
```diff
@@ -0,0 +1,182 @@
+            }
+
+            try {
+                Thread.sleep(delay_between_commands);
```
Is the delay only needed for the bridge or does the bulb electronics require a small delay. If the bulbs require a delay then does the delay need to span across bridges?

davidgraeff replied:
I recon the bridge can only handle a specific low amount of traffic in general, but we cannot control what other apps are sending whatsoever and the effort to realize bridge spanning delays for single bulbs is not worth it in my opinion

MikeJMajor replied:
After doing some experiments with a remote that hadn't been synced with my OH bulbs, it doesn't look like other bulbs signals interfere, so I don't think two bridges sending at the same time will be an issue.

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightCommunicationV3.java:
```diff
@@ -0,0 +1,645 @@
+            }
+            setPower(bulb, true, state);
+
+            sendQueue.add(messageBytes, bulb, (byte) 0, true);
```
This doesn't look right

davidgraeff replied:
For older bridges you were required to send the power on and then the actual command. I thought it wouldn't hurt if I'm doing it here as well, although the documentation doesn't require this.

MikeJMajor replied:
Sorry I should have been a bit clearer. Shouldn't the sendQueue.add be passing false for the repeats?

davidgraeff commented:
You're right the sendqueue.add argument has to be false

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3RGBW.java:
```diff
@@ -0,0 +1,267 @@
+
+            setPower(true, state);
+
+            sendQueue.queue(messageBytes, uidc(type_offset, CAT_SPEED_CHANGE), false);
```
For animation speed it's not being set to an absolute value so should you be discarding current speed commands on the queue? E.g. If I wrote a rule to turn on disco mode and set the speed to 5 by issuing 5 speed increases I would want all 5 sent.

davidgraeff replied:
i agree. i had a remote in mind where the user presses the speed change button rapidly but gets actual changes just every 100ms and the commands queue up and in the end the user is faced with a mich greater change than expected

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3RGBW.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    @Override
+    public void changeBrightness(int relative_brightness, MilightThingState state) {
```
changeBrightness isn't used for these bulbs since the brightness setting is absolute.

davidgraeff replied:
yup but changeBrightness is part of the interface and every bulb needs to implement this

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/AbstractBulbInterface.java:
```diff
@@ -0,0 +1,114 @@
+
+    public abstract void setSaturation(int value, MilightThingState state);
+
+    public abstract void changeBrightness(int relative_brightness, MilightThingState state);
```
Is changeBrightness needed in this interface as it doesn't seem to be called from the handler?

davidgraeff replied:
Thanks for spotting that. It is called from the handler now. It is not really necessary for the interface but it might make the life easier for script writers if the binding understands Increase/Decrease commands for brightness, don't you think so?

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3White.java:
```diff
@@ -0,0 +1,221 @@
+            state.colorTemperature = newPercent;
+        } else if (color_temp_relative < 0) {
+            logger.debug("milight: sendCooler");
+            int newPercent = state.colorTemperature - 10;
```
Is this missing a check to prevent newPercent going below zero?

davidgraeff replied:
Yes, the check was missing and is added now. Thanks for spotting that.

davidgraeff commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3RGBW.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Override
+    public void setPower(boolean on, MilightThingState state) {
+        byte command_on[] = { (byte) 0x42, (byte) 0x45, (byte) 0x47, (byte) 0x49, (byte) 0x4B };
```
Yes indeed. I made all those similar byte arrays final static now.

MikeJMajor commented:
I can't get discovery to work for RC1. Log gives the following error:

2017-02-07 21:02:47.061 [ERROR] [ht.internal.protocol.MilightDiscover] - Sending a discovery packet failed. Cannot assign requested address (sendto failed)

Bundle reports as:

204 | Active    |  80 | 2.1.0.ref-rc1         | Milight Binding

davidgraeff commented:
Could you check with this version again? I added the destination IP to the error message and avoid using disabled or loopback network interfaces now. The binding is sending on all interfaces, and uses IP4 and IP6, so there is a high chance that an error message will be caused by one of the interfaces but bridges will still be found on other interfaces.

[org.openhab.binding.milight_2.1.0.ref-rc2.jar.pdf](https://github.com/openhab/openhab2-addons/files/759315/org.openhab.binding.milight_2.1.0.ref-rc2.jar.pdf)

MikeJMajor commented:
I'll test again tomorrow, but looking at the code I'm not sure what it's trying to do.

Previously it looked like it sent the discovery packet to the broadcast address which was passed in from MilightBridgeDiscovery. Now it's sending to the addresses assigned to network interfaces, but that's not going to be the broadcast address so how does it get to the bridge?

davidgraeff commented:
Yes, I iterated over the IP addresses instead of broadcast addresses of the interfaces, good point and fixed.

MikeJMajor commented:
RC3 finds my v3 bridges, but they are discovered as v6 "Milight iBox..."

Should the binding be switching between v3 and v6 based on which discovery packets respond? Looking at the code it looks like its trying to do this based on the response message.

Due to it thinking its a v6 bridge it doesn't show any white bulbs, is it correct that an ibox cannot control white bulbs?

I can't seem to switch it into v3 from the paper UI, I get a server 500 error and this:

Status: ONLINE - CONFIGURATION_PENDING SESSION_INVALID

HIDE PROPERTIES
sessionid	00 00
version	6

I've tried stopping openhab and manually editing the JSON config file, but it still switches back to protocol version 6.

davidgraeff commented:
The protocol is absolutely different. The v6 things wouldn't be of any use for your setup. I have no idea what went wrong here, looks like your v3 bridge responses unexpected to discovery packets.

MikeJMajor commented:
Testing using a UDP tool from my phone they are responding as expected <IP>,<MAC>,

davidgraeff commented:
The check in the discovery class should recognize this as v3 if the length of the third string part is zero. I'll have a look later not at any pc rn.

MikeJMajor commented:
I've not been able to check, but my guess is that the following code should be passing in the length of the received packet or using packet.getData() when creating the string rather than using the entire buffer:

String[] msg = new String(buffer).split(",");

davidgraeff commented:
Yes that's the cause I think so too. I prepare another rc later.

davidgraeff commented:
[org.openhab.binding.milight_2.1.0.ref-rc4.jar.pdf](https://github.com/openhab/openhab2-addons/files/765516/org.openhab.binding.milight_2.1.0.ref-rc4.jar.pdf)

Changes:
I have modified the discovery and squashed all commits to a single one again. I think it is justified to have two bridge things here, so I implemented a MilightV3BridgeHandler/MilightV6BridgeHandler on top of an abstract bridge class. Especially because V6 does not need or use the discovery class after initial discovery anymore.

MikeJMajor commented:
With RC4 I can successfully control my rgbw and white lights using V3 bridges.

The bridge thing appears in the paper UI with a description, but is missing a name (the bold title)

Changing the repeats/delay from the paper UI causes OH to consume 100% and requires a reboot.

Somehow I managed to get it into a state in which it would not control the brightness of the rgbw bulbs when in white mode, but would work ok in colour mode.

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3RGBW.java:
```diff
@@ -0,0 +1,132 @@
+            byte messageBytes[] = new byte[] { 0x40, make_color(hue), 0x55 };
+            sendQueue.queueRepeatable(uidc(CAT_COLOR_SET), messageBytes);
+
+            if (brightness != -1) {
```
Brightness also needs to be applied when saturation is less than 50 (white mode)

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3RGBW.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    public void setBrightness(int value, MilightThingState state) {
```
setBrightness can be called using a decimal type so value needs to be prevented from going over 100

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3White.java:
```diff
@@ -0,0 +1,186 @@
+        if (relative_brightness == 0) {
+            return;
+        }
+        state.brightness = Math.max(state.brightness + relative_brightness, 0);
```
Need to prevent brightness going above 100

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/MilightV3White.java:
```diff
@@ -0,0 +1,186 @@
+    private final static byte prev_animation_mode[] = { 0x27, 0x00, 0x55 };
+
+    @Override
+    public void previousAnimationMode(MilightThingState state) {
```
This doesn't need an implementation since white bulbs don't have animation.

davidgraeff replied:
Are you really sure? Because other implementations have this implemented. Maybe there are only very simple animations like fading/flashing?

MikeJMajor replied:
There's nothing on the white bulb remote or on the easybulb app for animation for white bulbs

davidgraeff commented:
@MikeJMajor : Thanks for the review. I incorporated all suggested changes

MikeJMajor commented:
Do you have an rc5 build available for testing?

davidgraeff commented:
Not yet. I'll provide one in about 7 hours or so.

davidgraeff commented:
[org.openhab.binding.milight_2.1.0.ref-rc5.jar.pdf](https://github.com/openhab/openhab2-addons/files/782898/org.openhab.binding.milight_2.1.0.ref-rc5.jar.pdf)

MikeJMajor commented:
Brightness control for V3 bridge now works. I still have the problem of OH consuming 100% CPU after I have changed repeats/delay in the paper UI. Hopefully the following stack trace will help:


`Thread 1368: (state = IN_VM)
 - java.lang.Throwable.fillInStackTrace(int) @bci=0 (Compiled frame; information may be imprecise)
 - java.lang.Throwable.fillInStackTrace() @bci=16, line=783 (Compiled frame)
 - java.lang.Throwable.<init>() @bci=24, line=250 (Compiled frame)
 - java.lang.Exception.<init>() @bci=1, line=45 (Compiled frame)
 - java.lang.RuntimeException.<init>() @bci=1, line=51 (Compiled frame)
 - java.lang.IllegalStateException.<init>() @bci=1, line=44 (Compiled frame)
 - java.util.concurrent.LinkedBlockingQueue$Itr.remove() @bci=11, line=839 (Compiled frame)
 - org.openhab.binding.milight.internal.protocol.QueuedSend.remove_from_queue(int) @bci=47, line=178 (Compiled frame)
 - org.openhab.binding.milight.internal.protocol.QueuedSend.queueRepeatable(int, byte[][]) @bci=9, line=204 (Interpreted frame)
 - org.openhab.binding.milight.internal.protocol.MilightV3RGBW.setPower(boolean, org.openhab.binding.milight.internal.MilightThingState) @bci=40, line=54 (Interpreted frame)
 - org.openhab.binding.milight.internal.protocol.MilightV3RGBW.whiteMode(org.openhab.binding.milight.internal.MilightThingState) @bci=32, line=63 (Interpreted frame)
 - org.openhab.binding.milight.internal.protocol.MilightV3RGBW.setHSB(int, int, int, org.openhab.binding.milight.internal.MilightThingState) @bci=9, line=33 (Interpreted frame)
 - org.openhab.binding.milight.handler.MilightLedHandler.handleCommand(org.eclipse.smarthome.core.thing.ChannelUID, org.eclipse.smarthome.core.types.Command) @bci=584, line=88 (Interpreted frame)
 - org.eclipse.smarthome.core.thing.internal.ThingManager$4.call() @bci=12, line=372 (Interpreted frame)
 - org.eclipse.smarthome.core.thing.internal.ThingManager$4.call() @bci=1, line=1 (Interpreted frame)
 - org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call() @bci=11, line=179 (Compiled frame)
 - java.util.concurrent.FutureTask.run() @bci=42, line=266 (Compiled frame)
 - java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) @bci=95, line=1142 (Interpreted frame)
 - java.util.concurrent.ThreadPoolExecutor$Worker.run() @bci=5, line=617 (Interpreted frame)
 - java.lang.Thread.run() @bci=11, line=745 (Interpreted frame)`

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/QueuedSend.java:
```diff
@@ -0,0 +1,276 @@
+                while (repeat_remaining > 0) {
+                    datagramSocket.send(packet);
+                    --repeat_remaining;
+                    // logger.debug("Sent packet '{}' to bridge {}", s.toString(),
```
Is there a reason the debug logging is commented out?

Using RC5 my lights aren't switching as reliably as using my 2.0.0 binding modified to  repeat 4 times. It would be good to be able to verify the repeats/delay is working as expected.

davidgraeff replied:
It really spams the debugging console and all the other useful debug outputs get lost. It might be useful as a logger.trace() instead.

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/QueuedSend.java:
```diff
@@ -0,0 +1,276 @@
+            packet.setData(item.data);
+            try {
+                int repeat_remaining = item.repeatable ? repeat_commands : 1;
+                repeat_remaining = Math.min(repeat_remaining, 1);
```
I don't think the repeats work since this code looks like it will always set the number of repeats to 1:

repeat_remaining = Math.min(repeat_remaining, 1);

davidgraeff replied:
I've fixed that. Math.min/max still confuses me especially if it's late. I usually use bound(min, value, max) commands in other languages than java.

davidgraeff commented:
I implemented some details in the SendQueue slightly different now and hopefully don't get a IllegalStateException anymore. I'm not so familiar with multi thread promises of the LinkedBlockingQueue API for iteration, but let's give it a go.

[org.openhab.binding.milight_2.1.0.ref-rc6.jar.pdf](https://github.com/openhab/openhab2-addons/files/785498/org.openhab.binding.milight_2.1.0.ref-rc6.jar.pdf)

MikeJMajor commented:
With RC6 you can't control the bulbs after changing repeat/delay in paper UI, but it no longer consumes 100% CPU. A restart is required to get it working again.

Repeats still don't appear to be working. The repeat value is only set from the thingUpdated method, does that get called during startup or only when you change something in paper UI?

MikeJMajor commented:
The bulb objects are holding a reference to the queue, if for some reason during the call to thingUpdated the connectAndKeepAlive method is called a new queue object is created. Could the bulbs end up trying to use the old queue?

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/handler/AbstractMilightBridgeHandler.java:
```diff
@@ -0,0 +1,178 @@
+        // Create a new communication object if the user changed the bridge ID configuration.
+        String id_config = (String) thing.getConfiguration().get(MilightBindingConstants.CONFIG_ID);
+        if (id_config != null && !id_config.equals(bridgeid)) {
+            connectAndKeepAlive();
```
Is it possible that connectAndKeepAlive is getting called (is bridgeId initialised?) when I change repeats in the paper UI? This results in a new queue object being created, but are the bulb objects still referencing the old queue?

davidgraeff replied:
Yes, this binding was initially implemented in a way that **initialize()** is expected to be called only once and therefore only one sendqueue exists. All bulbs are referencing this very first object and never get updated with another. Additionally I didn't know that there is a **handleConfigurationUpdate()** method that needs to be overwritten which otherwise calls **initialize()** again and triggers the aforementioned problem.

davidgraeff commented:
Only when u change something. That's not right of course the values should be assigned at startup as well. Thingupdated seems to be broken if it makes the sendqueue not working anymore. I'll have a look later. I assume that the ip address or bridge ID is overwritten with wrong data.

davidgraeff commented:
[org.openhab.binding.milight_2.1.0.ref-rc7.jar.pdf](https://github.com/openhab/openhab2-addons/files/789375/org.openhab.binding.milight_2.1.0.ref-rc7.jar.pdf)

Repeat value and delay time are no applied at start time and on configuration change.

davidgraeff commented:
Changes:

* rc9: QueuedSend repaired. Repeated OFFLINE/ONLINE status fixed.
* rc8: Repeat value and delay time are now applied at start time and on configuration change. Implemented handleConfigurationUpdate to avoid dispose/initialize process. Improve thing status change depending on the bridge status. Always keep the same sendQueue object in the bridge.

[org.openhab.binding.milight_2.1.0.ref-rc9.jar.pdf](https://github.com/openhab/openhab2-addons/files/800962/org.openhab.binding.milight_2.1.0.ref-rc9.jar.pdf)

colinfrei commented on addons/binding/org.openhab.binding.milight/README.md:
```diff
@@ -1,92 +1,137 @@
 
 The Thing configuration for the bridge uses the following syntax
-Bridge milight:bridge:<mac address of bridge> [ ADDR="<IP-Address of bridge>", PORT=<listening port> ]
+Bridge milight:bridge:<mac address of bridge> [ ADDR="<IP-Address of bridge>", CONFIG_PROTOCOL_VERSION=3/6 ID=<mac> ]
```
Shouldn't this be
Bridge milight:bridgeV6:<mac address of bridge>
?

Same in the example above. I'm not sure if CONFIG_PROTOCOL_VERSION is still used then?

davidgraeff replied:
you're right. This detail changed during development and the readme wasn't updated. Good catch.

MikeJMajor commented:
With RC9 changing settings in paper UI looks like its working and individual control of bulbs with V3 bridges looks good.

The bulb control seems much more reliable so it looks like the repeats are working.

However, there are problems with the wrong bulbs changing. E.g. turn on bulb 1, bulb 2 and bulb 3. Attempting to change bulb 2 to colour mode sometimes switches bulb 3 to colour (bulb 2 is still white), but then turning off bulb 2 works OK. I've also seen both bulb 2 and bulb 3 change colour at the same time.

MikeJMajor commented:
I think the queue logic is trying to be too clever and incorrectly removing commands.

When a light is set to a colour the commands you want to send are: power, colour, power, brightness. 

I'm guessing when it goes wrong the first power is still on the queue and gets removed so it sends: colour, power, brightness.

m3tatai commented:
Hello @MikeJMajor,
what kind of command you are using to change color and brightness?

davidgraeff commented:
You're right. The second power is removed. The queue logic needs a dependency think apparently

davidgraeff commented:
Could you test with this version?:

[org.openhab.binding.milight_2.1.0.ref-rc10.jar.pdf](https://github.com/openhab/openhab2-addons/files/803604/org.openhab.binding.milight_2.1.0.ref-rc10.jar.pdf)

I have implemented chainable QueuedSend items, with a chain of commands is defined as being processed in total or not at all (if removed by a same category command).

Cheers, David

MikeJMajor commented on addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/protocol/QueuedSend.java:
```diff
@@ -0,0 +1,344 @@
+            }
+
+            if (item == null || item.unique_command_id == QueueItem.INVALID) {
+                continue;
```
Only the first item in a chain is invalidated so when you call continue it will send the second command.

Should you be setting item to null if the start of the chain is invalid?

davidgraeff replied:
You are right, has been fixed with the next commit.

MikeJMajor commented:
With RC10 I think I spotted the wrong bulb changing colour once, but couldn't reproduce it.  This might be due to the issue I've mentioned with invalidating a chained queue item.

I need to do a bit more testing with white bulbs as I'm not sure the brightness setting looked correct.

davidgraeff commented:
If it's almost perfect can we give it a go? A lot of people are waiting for the ibox  support and I only expect small fixes solvable in separate issues to be left. 

Cheers David

MikeJMajor commented:
Do you agree there's an issue with a chained queue item still sending commands after it's been set to invalid, I.e. it skips the first command in the chain but still sends the rest?

The queue logic doesn't look right for setting the brightness for white bulbs, e.g if the bulb is off and you set the brightness to 50 followed immediately by 10 then it will attempt to remove the 50 chain. However only the 50 chain has the set full command so the internal state will be out of sync with the bulb. I don't think you can remove items from the queue when you need to rely on the internal state.

I can't say if these issues are going to cause other v3 user issues.

davidgraeff commented:
V3 has a lot of limitations. The white bulbs with only supporting a relative brightness adjustment is one of the annoying things. We are out of sync as soon as someone use an external remote. But the current mechanism of the send queue doesn't allow your scenario at least. If there is a 50 brightness command chain in the queue it will be processed entirely. If there is a 10 command following it will invalidate the first chain command of the 50 chain, that is correct. But that doesn't change anything because the chain is either already executing or the 50 chain will be ignored entirely because the first command is invalid. Does that make sense? 

There might be implementation bugs though. I didn't setup a test case yet.

MikeJMajor commented:
Yes it makes sense, but for the case when the 50 chain is ignored entirely, the internal state is still set to 50.   So going from 50 to 51 would  be seen as a brightness increase of 1 rather than 51 (since the 50 hasn't actually been actioned)

I'm only mentioning it since when I was playing around with a brightness slider I noticed the brightness of the bulb seemed a lot brighter than it should for the position of the slider, which is why I thought something wasn't quite right.

davidgraeff commented:
What might solve this, is applying the final brightness value only after the chain has been processed. The SendQueue items could have a callback for that. I wonder if that makes the design over complicated.

MikeJMajor commented:
For the white brightness and colour temperature  percentage setting  you might as well  just send all the commands (rather than trying to remove queued commands) since it already uses it's internal state to work out what commands you need to send, e.g if you sent 50 followed by 50 it would only send commands for the first 50

davidgraeff commented:
Yes all relative commands will be allowed to queue up then which is basically all non repeatable commands.

davidgraeff commented:
[org.openhab.binding.milight_2.1.0.ref-rc11.jar.pdf](https://github.com/openhab/openhab2-addons/files/810363/org.openhab.binding.milight_2.1.0.ref-rc11.jar.pdf)

koimotion commented:
David, I am not sure if this is related to queues or your binding but I tried rc11 with colour picker control of rgbww lights using my android phone. I get one shot at hue on the colour ring and this works as do the saturation and brightness sliders inside the ring. Other hue selections do nothing on the android or change light colour. I have to exit the colour picker and go back in. I then get one more chance to change hue and it "locks up" again.The sliders do work.

davidgraeff commented:
Please double check with the paperui. It is a new v6 bulb?

davidgraeff commented:
@kaikreuzer: Please pull this merge request. There is an increasing demand for the new supported bridge in this version and all I can do in the forum is pointing people to the provided jar in this thread. There are non outstanding issues I am aware of and I prefer to fix any later issues in smaller pull requests.

Thanks, David

koimotion commented:
The lights are RGBWW/v6 and I did some further tests and habpanel and paper ui work fine. I used the colour picker in the basic ui and found that initial clicks were not changing colour. Waiting a while worked with each click on the colour wheel, so I went back to the colour picker in the openhab Android app and found that waiting and being very careful re picking the slider dots on the hue wheel and saturation/brightness bars did work. The issues look like a combination of Android app selection "hard to use" and system responsiveness. The binding looks good to release.

davidgraeff commented:
@koimotion Your observations are interesting though. The way the binding works now is:
* Adding your command to a queue.
* If your command is changing something like brightness that an older command in the queue already wanted to change -> make the old command invalid and use the new one instead. 

The queue will be processed in the background and commands are constantly send to the bridge if there are any with a delay of ~100ms (the bridge is quite slow).

If you slide with your finger over the colorwheel, a lot of commands are issued. But only a few (the latest every 100ms) will ever be send to the bridge. Your described behavior sounds like there is something going wrong somewhere. I will investigate this after the merge.

Could you please set your OH2 log level to debug and append your log to the community forum thread?

Cheers, David

davidgraeff commented:
I will not have WiFi the next weeks so I'm not able to do the review process fixes. Can anybody else do this?

kaikreuzer commented:
@koimotion Thanks for helping, I very much appreciate any help on the review process.
Please note though, that with adding an "approval" here on Github, you declare that you have checked the PR against all [our code guidelines](http://docs.openhab.org/developers/development/guidelines.html). Can you confirm this?

koimotion commented:
@kaikreuzer  Thanks for your feedback and following your comment re code guidelines I have reviewed again all the code in the pull request. I am new to openHAB and have limited Java experience. However, I can say that I could not find any cases where the coding guidelines have not been followed.

micropet commented:
Danke David für Deine tolle arbeit.
Meine v6 Bridge funktioniert tadellos.
Die Verbindungsabbrüche sind aber noch da:

2017-03-27 17:58:27.973 [hingStatusInfoChangedEvent] - 'milight:bridgeV3:ACCF236D3F2C' changed from ONLINE to OFFLINE
2017-03-27 17:58:27.973 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:0' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.973 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:1' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.974 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:4' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.974 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:2' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.974 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:3' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.974 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:0' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.974 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:1' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.975 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:4' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.975 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:2' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:27.975 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:3' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
2017-03-27 17:58:28.979 [ItemStateChangedEvent     ] - D25_ke_Pir changed from 1 to 0
2017-03-27 17:58:29.022 [ItemStateChangedEvent     ] - D25_ku_DS_Temp1 changed from 25.50 to 25.44
2017-03-27 17:58:29.368 [ItemStateChangedEvent     ] - ESP_32_DS_Temp changed from 28.56 to 28.62
2017-03-27 17:58:29.626 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:0' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.626 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:1' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.627 [hingStatusInfoChangedEvent] - 'milight:bridgeV3:ACCF236D3F2C' changed from OFFLINE to ONLINE
2017-03-27 17:58:29.627 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:4' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.627 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:2' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.628 [hingStatusInfoChangedEvent] - 'milight:whiteLed:ACCF236D3F2C:3' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.628 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:0' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.628 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:1' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.629 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:4' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.629 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:2' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2017-03-27 17:58:29.629 [hingStatusInfoChangedEvent] - 'milight:rgbLed:ACCF236D3F2C:3' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
